SwiftGit2/Repository.swift

```diff
@@ -558,4 +558,137 @@ final public class Repository {
 		let iterator = CommitIterator(repo: self, root: branch.oid.oid)
 		return iterator
 	}
+
+	// MARK: - Diffs
+
+	public func diff(for commit: Commit) -> Result<[Diff.Delta], NSError> {
+		return self
+			.commit(with: commit.oid)
+			.flatMap { baseCommit in
+				return self.tree(from: baseCommit)
+			}
+			.flatMap { baseTree in
+				guard !commit.parents.isEmpty else {
+					// Initial commit in a repository
+					return self.diff(fromTree: nil, toTree: baseTree)
+				}
+
+				var mergeDiff: Result<OpaquePointer?, NSError> = .success(nil)
+				for parent in commit.parents {
+					let diff = self
+						.commit(with: parent.oid)
+						.flatMap { parentCommit in
+							return self.tree(from: parentCommit)
+						}
+						.flatMap { parentTree in
+							return self.diff(fromTree: parentTree, toTree: baseTree)
+						}
+					mergeDiff = mergeDiff
+						.fanout(diff)
+						.flatMap { (mergeDiff, diff) in
+							guard let mergeDiff = mergeDiff else {
+								return .success(diff)
+							}
+							let mergeResult = git_diff_merge(mergeDiff, diff)
+							guard mergeResult == GIT_OK.rawValue else {
+								return .failure(NSError(gitError: mergeResult, pointOfFailure: "git_diff_merge"))
+							}
+							return .success(mergeDiff)
+						}
+				}
+				return .success(mergeDiff.value!!)
+			}
+			.flatMap { diffResult in
+				return self.processDiffDeltas(diffResult)
+			}
+	}
+
+	private func commit(with oid: OID) -> Result<OpaquePointer, NSError> {
+		var unsafeBaseCommit: OpaquePointer? = nil
+		let unsafeBaseOid = UnsafeMutablePointer<git_oid>.allocate(capacity: 1)
+		git_oid_fromstr(unsafeBaseOid, oid.description)
+		let lookupBaseGitResult = git_commit_lookup(&unsafeBaseCommit, self.pointer, unsafeBaseOid)
+		guard lookupBaseGitResult == GIT_OK.rawValue, let unwrapBaseCommit = unsafeBaseCommit else {
+			return Result.failure(NSError(gitError: lookupBaseGitResult, pointOfFailure: "git_commit_lookup"))
+		}
+		git_commit_free(unsafeBaseCommit)
+
+		return Result.success(unwrapBaseCommit)
```

Comment:
I don't think it's safe to use `unwrapBaseCommit` here after freeing `unsafeBaseCommit`. I think this should switch to use the same callback style as the `remoteLookup` method.

Reply:
I was working on this when I discovered the already existing `withGitObject`… what do you think about replacing e.g. `.commit(with: parent.oid)` with `.withGitObject(parent.oid, type: GIT_OBJ_COMMIT)` and killing the `commit(with:)` altogether? The tests still pass.

Reply:
Yeah, that sounds good! 👍 You can tell it's been a while since I've really worked with this code. 😅

Reply:
No worries. There's actually a handful of helpers I never noticed before, so I'll convert things to use those.